Add `onbegin` event in `SVGAnimationElement.idl`

Ahmad-S792 opened the pull request:
#### db45fa396ca71e63e7879629e810755c7c818f22
<pre>
Add `onbegin` event in `SVGAnimationElement.idl`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275669">https://bugs.webkit.org/show_bug.cgi?id=275669</a>
<a href="https://rdar.apple.com/problem/130609424">rdar://problem/130609424</a>

Reviewed by Said Abou-Hallawa.

This patch adds &apos;onbegin&apos; event in &apos;SVGAnimationElement&apos; IDL interface
as per web specification [1]:

[1] <a href="https://svgwg.org/specs/animations/#InterfaceSVGAnimationElement">https://svgwg.org/specs/animations/#InterfaceSVGAnimationElement</a>

* Source/WebCore/dom/EventNames.json:
* Source/WebCore/svg/SVGAnimationElement.idl:
* LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt:

&gt; Rebaselines:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:

Canonical link: <a href="https://commits.webkit.org/299724@main">https://commits.webkit.org/299724@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c422b189ed332095294bbdbc18c857c3b2890bc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39655 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126294 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72035 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d8cc8a48-7757-4d20-a33c-13c3eaef4c8e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121838 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91099 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60411 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bdfc2711-bdb9-40ea-8da2-7b4d9039ce41) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32229 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107565 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71655 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d683b5c9-f286-41a3-8277-466de610e40f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31260 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25671 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69927 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101704 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25859 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129208 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46882 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35545 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99717 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103754 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25281 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45015 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23030 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43503 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46744 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52450 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46210 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49559 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47896 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt:
```diff
@@ -1631,5 +1646,188 @@ PASS SVGElement interface: objects.view must inherit property "ownerSVGElement"
+PASS SVGAnimationElement interface: attribute targetElement
+PASS SVGAnimationElement interface: attribute onbegin
+FAIL SVGAnimationElement interface: attribute onend assert_true: The prototype object must have a property "onend" expected true got false
+FAIL SVGAnimationElement interface: attribute onrepeat assert_true: The prototype object must have a property "onrepeat" expected true got false
```
Why do not you add 
```
attribute EventHandler onend;
attribute EventHandler onrepeat;
```
to the IDL to match the spec and to fix these failures?

Ahmad-S792 replied:
We need to do full `repeatEvent` implementation first - which is this PR - https://github.com/WebKit/WebKit/pull/29658 and for `onend`, it expose another bug and lead to WPT failure, which needs further investigation. So I am doing minimum, which is not regressing anything.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt:
```diff
@@ -1631,5 +1646,188 @@ PASS SVGElement interface: objects.view must inherit property "ownerSVGElement"
+PASS SVGAnimationElement interface: objects.set must inherit property "targetElement" with the proper type
+PASS SVGAnimationElement interface: objects.set must inherit property "onbegin" with the proper type
+FAIL SVGAnimationElement interface: objects.set must inherit property "onend" with the proper type assert_inherits: property "onend" not found in prototype chain
+FAIL SVGAnimationElement interface: objects.set must inherit property "onrepeat" with the proper type assert_inherits: property "onrepeat" not found in prototype chain
```
Ditto.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt:
```diff
@@ -1631,5 +1646,188 @@ PASS SVGElement interface: objects.view must inherit property "ownerSVGElement"
+PASS SVGAnimationElement interface: objects.animateTransform must inherit property "targetElement" with the proper type
+PASS SVGAnimationElement interface: objects.animateTransform must inherit property "onbegin" with the proper type
+FAIL SVGAnimationElement interface: objects.animateTransform must inherit property "onend" with the proper type assert_inherits: property "onend" not found in prototype chain
+FAIL SVGAnimationElement interface: objects.animateTransform must inherit property "onrepeat" with the proper type assert_inherits: property "onrepeat" not found in prototype chain
```
Ditto.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window.js:
```diff
@@ -171,40 +154,24 @@ idl_test(
       SVGViewElement: ['objects.view'],
       SVGScriptElement: ['objects.script'],
       SVGAnimateElement: ['objects.animate'],
+      SVGDiscardElement: ['objects.discard'],
```
Do we have a bug for adding the SVG `<discard>` Element? If we do not have one, please file a new bug.

Ahmad-S792 replied:
I can file but Blink had the support but removed it - https://groups.google.com/a/chromium.org/g/blink-dev/c/jUJG_CFqEvE/m/bc5_tXo8BAAJ and Gecko only had `Intent to Prototype` - https://groups.google.com/a/mozilla.org/g/dev-platform/c/U4nufgUTdAk

shallawa replied:
Then I think we should remove this line for now. This file has been out of sync for sometime. And I think it is okay to it have out of sync for the `<discard>` object since the plan for it is unclear.

I think removing this line will remove many failures from the expected file.

shallawa commented on LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window.js:
```diff
@@ -28,41 +28,24 @@ const elements = [
   'script',
   'animate',
   'set',
+  'discard',
```
Please remove this line.

Ahmad-S792 replied:
I synced it from WPT - if we need to delete, we should do it upstream and if I delete it from local copy, upon next WPT sync, it will be added back.

shallawa replied:
So what is the purpose of this PR then?

1. We can't add `onrepeat` because we have to implement `repeatEvent` first
2. We can't add `onend` because it exposes a bug that we need to fix first
3. We can't assume `<discard>` will be considered because Blink has dropped it and Gecko has the intent to prototype it.
4. We can't remove `<discard>` from idlharness.window.js because we will have to upstream it.

Ahmad-S792 replied:
It is to add `onbegin`, which we don't add yet.

shallawa replied:
Yes but you are introducing many other failures which are not clear to me if they are real failures or not. For me the _expected.txt_ file is the truth. So when I see a failure like 
```
FAIL SVGDiscardElement interface: existence and properties of interface object assert_own_property: self does not have own property "SVGDiscardElement" expected property "SVGDiscardElement" missing
```
It tells me that we have a bug by not implementing `SVGDiscardElement` which is not true.

So if we add this line to the _expected.txt_ file, it will not be the truth for me anymore.

shallawa replied:
You can add `onbegin` without syncing the WPT.

Ahmad-S792 replied:
OK, I will remove updated test. :-)

Ahmad-S792 replied:
Done.

Commit:
Add `onbegin` event in `SVGAnimationElement.idl`
https://bugs.webkit.org/show_bug.cgi?id=275669
rdar://problem/130609424

Reviewed by Said Abou-Hallawa.

This patch adds 'onbegin' event in 'SVGAnimationElement' IDL interface
as per web specification [1]:

[1] https://svgwg.org/specs/animations/#InterfaceSVGAnimationElement

* Source/WebCore/dom/EventNames.json:
* Source/WebCore/svg/SVGAnimationElement.idl:
* LayoutTests/imported/w3c/web-platform-tests/svg/idlharness.window-expected.txt:

> Rebaselines:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative-expected.txt:

Canonical link: https://commits.webkit.org/299724@main
